plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/editor/context/project/EncoderServer.kt

```diff
@@ -0,0 +1,263 @@
+                if (validateHash(serverContent.hashes?.first(), HttpRequests.request(serverContent.url).readBytes(null))) {
+                    downloadFromRemote(serverContent.url, zipFilePath)
+                    unzipFile(zipFilePath, cachePath)
+                }
+            }
+        } catch (e: Exception) {
+            logger.warn { "error downloading artifacts ${e.stackTraceToString()}" }
+        }
+    }
+
+    private fun tryDeleteOldArtifacts(filenames: List<String>) {
+        try {
+            filenames.forEach { filename ->
+                run {
+                    if (filename.contains("qserver")) {
+                        val parts = filename.split("-")
+                        val version = if (parts.size > 1) parts[1] else null
+                        if (version != null && version != "${manifestManager.currentVersion}.zip") {
+                            Files.deleteIfExists(cachePath.resolve(filename))
+                            Files.deleteIfExists(cachePath.resolve("qserver"))
+                        }
+                    }
+                }
+            }
+        } catch (e: Exception) {
+            logger.warn { "error deleting old artifacts $e.stackTraceToString()" }
+        }
+    }
+
+    fun validateHash(expectedHash: String?, input: ByteArray): Boolean {
+        if (expectedHash == null) { return false }
+        val sha384 = DigestUtils.sha384Hex(input)
+        val isValid = ("sha384:$sha384") == expectedHash
+        if (!isValid) {
+            logger.warn { "failed validating hash for artifacts $expectedHash" }
+        }
+        return isValid
+    }
+
+    private fun makeFileExecutable(filePath: Path) {
+        val permissions = setOf(
+            PosixFilePermission.OWNER_READ,
+            PosixFilePermission.OWNER_WRITE,
+            PosixFilePermission.OWNER_EXECUTE,
+            PosixFilePermission.GROUP_READ,
+            PosixFilePermission.GROUP_EXECUTE,
+            PosixFilePermission.OTHERS_READ,
+            PosixFilePermission.OTHERS_EXECUTE,
+        )
+        Files.setPosixFilePermissions(filePath, permissions)
+    }
+
+    private fun unzipFile(zipFilePath: Path, destDir: Path) {
+        if (!zipFilePath.exists()) return
+        try {
+            val zipFile = ZipFile(zipFilePath.toFile())
+            zipFile.use { file ->
+                file.entries().asSequence()
+                    .filterNot { it.isDirectory }
+                    .map { zipEntry ->
+                        val destPath = destDir.resolve(zipEntry.name)
+                        destPath.createParentDirectories()
+                        FileOutputStream(destPath.toFile()).use { targetFile ->
+                            zipFile.getInputStream(zipEntry).copyTo(targetFile)
+                        }
+                    }.toList()
+            }
+        } catch (e: Exception) {
+            logger.warn { "error while unzipping project context artifact: ${e.message}" }
+        }
+    }
```

Comment:
btw zoe what's difference between this and https://github.com/aws/aws-toolkit-jetbrains/blob/main/plugins/amazonq/codetransform/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codemodernizer/utils/CodeTransformFileUtils.kt#L76-L91

Reply:
I have a SIM to dedupe these two functions, need to sort out the dependency